Partial support for namespace scope variables.

AlexVlx opened the pull request:
This adds incomplete support for generalised C++ language namespace scope variables, relying on ELF introspection and relevant HSART interfaces. At the moment there is an issue with observing accelerator writes to a global correctly, and support for shared objects using a global defined in main is unimplemented. The latter would require quite a bit of backporting from HC2, which unless proven to be utterly necessary can be eschewed.

AlexVlx committed:
This adds incomplete support for generalised C++ language namespace scope variables, relying on ELF introspection and relevant HSART interfaces. At the moment there is an issue with observing accelerator writes to a global correctly, and support for shared objects using a global defined in main is unimplemented. The latter would require quite a bit of backporting from HC2, which unless proven to be utterly necessary can be eschewed.

AlexVlx committed:
Add missing files to prior commit.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -271,6 +271,143 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
+        const auto undefined_symbols = copy_names_of_undefined_symbols(
+            symbol_section_accessor{reader, code_object_dynsym});
+
+        for (auto&& x : undefined_symbols) {
```
Maybe we should check whether `x` has already been defined for this `agent` and if it has already been defined, then skip to the next undefined symbols?

AlexVlx replied:
We could, but this would only be beneficial if we had multiple code objects per agent (otherwise having multiple unmerged symbols with the same name in the same code object / TU would be an ODR violation / not possible), which I do not believe we do at this point in time. The underlying assumption here is that the mapping from code-object undefined symbol to loader allocated host symbol is bijective. However, I may be missing something about how "executables" is managed at the moment...

scchan replied:
If an application is linked to a shared library with GPU code then it's possible that an agent will have more than one code objects (i.e. one from the application's objects and another from the shared library).

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -271,6 +271,143 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
+
+            void* p = nullptr;
+            hsa_amd_memory_lock(
+                reinterpret_cast<void*>(tmp.first), tmp.second, &agent, 1, &p);
```
Should we check whether the host variable has already been "pinned" previously?

AlexVlx replied:
We could, but this would require either using a slightly more complex data structure or O(n) searching on each lock. If the mapping is bijective, this appears unnecessary, for there can be no scenario in which multiple different symbols try to lock the same loader allocated variable in the executable's image. If we were to have multiple code-objects containing code which references the same global, and which are loaded during the same execution, this may be an optimisation worth pursuing. At the moment we do not support it, but a scenario for it would be both the executable and shared object(s) referencing a global that is defined in a shared object.

scchan replied:
Any reason why we don't support multiple objects referencing the same global?  It seems to be that your algorithm should just work and we just need verify whether ROCR is able to handle double locking/unlocking the same global.  If ROCR couldn't handle it properly than hcc RT would need to manage the memory lock/unlock more carefully.

scchan commented on tests/Unit/NamespaceScopeVariables/global.cpp:
```diff
@@ -0,0 +1,53 @@
+        return EXIT_FAILURE;
+    }
+
+    #if false // 10/09/2017 - writes to globals are not correctly observed.
```
any idea why this doesn't work

AlexVlx replied:
@scchan not at this moment, and unless there is a strong need for that now (I do not believe that is the case, but I may be wrong) I would punt on it. Mainly due to time constraints and it working in HC2. It is your call though.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_add_partial_support_for_namespace_scope_variables

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_add_partial_support_for_namespace_scope_variables

# Conflicts:
#	lib/hsa/mcwamp_hsa.cpp

whchung reviewed (dismissed):
to begin with, I think this PR deserves a better 1-line summary? "Add missing files to prior commit." is misleading.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_add_partial_support_for_namespace_scope_variables

AlexVlx committed:
Cache global symbols after locking so as to prevent multiple calls to
hsa_amd_memory_lock / hsa_amd_memory_unlock against the same address.
Correctly retrieve global symbols from both the current process and
linked in shared objects.

AlexVlx committed:
Cache global symbols after locking so as to prevent multiple calls to
hsa_amd_memory_lock / hsa_amd_memory_unlock against the same address.
Correctly retrieve global symbols from both the current process and
linked in shared objects.

AlexVlx committed:
Merge remote-tracking branch 'origin/feature_add_partial_support_for_namespace_scope_variables' into feature_add_partial_support_for_namespace_scope_variables

# Conflicts:
#	lib/hsa/mcwamp_hsa.cpp

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -271,6 +272,207 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
+                        "Global symbol: " + x + " is undefined."};
+                }
+
+                if (globals.find(x) == globals.cend()) {
```
I don't see `x` being added to `globals` or maybe I'm missing something?

AlexVlx replied:
I guess I was mentally transporting the electrons:) Thank you, fixed.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -271,6 +272,207 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
+
+                if (globals.find(x) == globals.cend()) {
+                    static mutex mtx;
+                    lock_guard<mutex> lck{mtx};
```
I presume this mutex is guarding `globals` for the insertion of `x` so shouldn't the lock comes before the 2nd `find(x)`?

AlexVlx replied:
Not quite; `find` is `const`, therefore thread safe, doing a const read under a lock would be a pessimisation. If we really did not find it (i.e. nobody added it in between our second search) then we actually go and lock mtx. Since we already reserved the maximum possible amount of memory, the insertion would not invalidate any iterators so the return value of find or cend would not be rendered bogus by another thread doing an insertion. Bad things will happen if some detached thread somewhere actually tries to do this, since it may be running after main and static destruction have returned, but that's a different story altogether and a programmer error.

scchan replied:
Don't you need to prevent another thread inserting the same `x` between the 2nd `find(x)` and the lock?

AlexVlx replied:
I was about to say that that is benign since it would only yield multiple calls to hsa_amd_memory_lock, but that's incorrect since the destruction of the temporary would call hsa_amd_memory_unlock on possibly the same, already inserted p. Thus you are correct; fixed it, thank you.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_add_partial_support_for_namespace_scope_variables

AlexVlx committed:
This corrects an oversight which resulted in locked globals never being
cached.

AlexVlx committed:
This removes a potential multi-unlock scenario.

scchan reviewed (approved):
LGTM

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_add_partial_support_for_namespace_scope_variables

scchan commented:
@whchung do you have other concerns regarding this PR?

whchung commented:
@scchan please go ahead and merge this PR
